### PR TITLE
rewrite of libgamename shared library to externalize the list of arcade games

### PR DESCRIPTION
**Pull request by Macs75**

### CODE REVIEW

File: src/tweaks/tools.h [hunk 1]
```diff
@@ -127,8 +127,14 @@ void tool_generateCueFiles(void *pt)
                      "/mnt/SDCARD/.tmp_update/script/cue_gen.sh");
 }
 
+void tool_buildShortRomGameList(void *pt)
+{
+    _runCommandPopup(tools_short_names[6],
+                     "./bin/gameNameList /mnt/SDCARD");
```
Comment (Aemiii91) [inline src/tweaks/tools.h:133]: A better location could be `/mnt/SDCARD/BIOS/arcade_lists` or something like that

File: Makefile [hunk 2]
```diff
@@ -84,6 +84,9 @@ $(CACHE)/.setup:
 		$(SRC_DIR)/randomGamePicker \
 		$(SRC_DIR)/easter \
 		-depth -type d -name res -exec cp -r {}/. $(BUILD_DIR)/.tmp_update/res/ \;
+	@find \
+	    $(SRC_DIR)/libgamename \
+		-depth -type d -name res -exec cp -r {}/. $(BUILD_DIR)/ \;
```
Comment (Aemiii91) [inline Makefile:89]: Not a fan of this location

**Review by Aemiii91 — CHANGES_REQUESTED** [review]
To match the code style, function names should be this format: `namespacePrefix_functionNameIsHere` instead of snake case (I know it's very mixed in our source, but I strive to stick with this format)

File: Makefile (hunk 2 above)
Comment (Macs75) [reply Makefile:89]: This is the txt file containing the full list of games extracted from the various emulators. It is modifiable by the users if the want add games or change the titles to display. I thought in the root was quite evident to see and easily updatable, but if you have a better proposal where to put it I will change.

File: src/tweaks/tools.h (hunk 1 above)
Comment (Macs75) [reply src/tweaks/tools.h:133]: are you referring where to put the list or the tool?

Comment (Aemiii91) [reply src/tweaks/tools.h:133]: The list

File: Makefile (hunk 2 above)
Comment (Aemiii91) [reply Makefile:89]: See my other comment

**Comment by anzz1** [comment]
As the whole codebase is currently in C, wouldn't adding the C++ STL just for this one lib increase the codebase size quite significantly? Also this firmware being essentially kernel code, are you sure you should introduce STL to the kernel, especially as this simple functionality could be easily made in plain C? 

While slightly beside the point, there's a reason why both Linux/Windows use C for the kernel code and not C++. For example, when building windows drivers with the [/kernel switch](https://learn.microsoft.com/en-us/cpp/build/reference/kernel-create-kernel-mode-binary?view=msvc-170), it disallows the use of STL and most of C++'s functions like exception handling and polymorphism. Linux kernel also doesn't use C++, [here's some comments from Linus Torvalds regarding that](https://itwire.com/business-it-news/open-source/rust-support-in-linux-may-be-possible-by-5-14-release-torvalds.html).

I don't claim to be familiar with the hardware yet (I'm receiving my MMP only tomorrow) nor ARM in general, I'm just advising caution to be really 100% sure about what you're doing here.

**Comment by Macs75** [comment]
are you referring to gamename.c -> libgamename.so ?
that c++ library already exists in the original miyoo and it's and is used by MainUi , I only rewritten it keeping the original
signature to do something different ( better ). It has to remain C++ or MainUi would not be able to access it (due to the name mangling used by C++  when generating the binary)
Now I use it also from the code of the new tool. Actually my rewritten library is 320k instead of the original 1,5Mb

**Comment by anzz1** [comment]
:+1: Sounds good then ! I didn't realize that the code is replacing something outside the repo (the original .so file).

1,5Mb original binary sounds heavy and probably is using STL anyway. Is the original firmware package available somewhere or do I have to extract it from the device? I presume you decompiled the original .so to rewrite it's functionality? Just interested to look at this stuff when I get my hands on the device.

It's actually possible to compile kernelmode code in C++ too, it's just that STL constructs like std::map aren't available, and exception handlers need to be disabled.

> It has to remain C++ or MainUi would not be able to access it (due to the name mangling used by C++ when generating the binary)

This isn't actually true though, you can code in C/C++ and export a function with whatever name, regardless of name mangling. This can be done with a linker definition `.def` file, or with a `/EXPORT` directive.
You can inline such a thing in the code without changing project settings like this:
```c
// export C func as C++
long __cdecl MyFunc(long i) {
  #pragma comment(linker, "/EXPORT:_MyFunc@4=_MyFunc")

  return i;
}

// export C++ func as C
long __cdecl MyFunc(long i) {
  #pragma comment(linker, "/EXPORT:_MyFunc=_MyFunc@4")

  return i;
}
```
You can obviously have it have any name with the `/EXPORT:whatever=yourfunc` directive.

**Comment by Macs75** [comment]
> 1,5Mb original binary sounds heavy and probably is using STL anyway

It is a really simple library. All the space is taken by 2 huge array of data containing the association of romname and title for English and Chinese language. The library is on the sd card itself like the other programs. 
Good to know about the #pragma comment!

**Comment by anzz1** [comment]
> > 1,5Mb original binary sounds heavy and probably is using STL anyway
> 
> It is a really simple library. All the space is taken by 2 huge array of data containing the association of romname and title for English and Chinese language. The library is on the sd card itself like the other programs. Good to know about the #pragma comment!

Oh yeah, you're correct. It wasn't using STL, infact the original doesn't use any C++ even if it exports the `GetGameName` function in a C++ mangled style. So it probably means it was originally compiled as .cpp and not .c , maybe by a mistake or a because of a requirement regarding the name mangling.

It only imports the `strcmp()` , `strncmp()`, `__cxa_finalize()` functions from glibc and nothing else. No C++ STL. It exports the `s_mamedb_en` and `s_mamedb_ch` data sections for the names and a single function, `GetGameName()`.

Here's the full function:
```c
char *__fastcall GetGameName(const char *lang, const char *name)
{
  int i; // [sp+Ch] [bp-10h]
  int j; // [sp+Ch] [bp-10h]

  if ( !name )
    return 0;
  if ( !strncmp(lang, "ch.lang", 7u) )
  {
    for ( i = 0; i < 11047; ++i )
    {
      if ( (&s_mamedb_ch)[2 * i] && !strcmp(name, (&s_mamedb_ch)[2 * i]) )
        return (&s_mamedb_ch)[2 * i + 1];
    }
  }
  for ( j = 0; j < 13883; ++j )
  {
    if ( s_mamedb_en[2 * j] && !strcmp(name, s_mamedb_en[2 * j]) )
      return s_mamedb_en[2 * j + 1];
  }
  return 0;
}
```
Indeed, very simple. While using a simple loop with strcmp isn't ideal, a fast hashmap would be better, it's okay performance wise. Having the names in a static data section of the library means that it helps with the very limited RAM available, as there are no dynamic memory allocations and thus no need to swap and r&w any memory at any point to the SD card.

I don't think the C++ STL is included on any of the system components and I'm not sure it should be done here either.
In any case, I'd carefully test and benchmark the performance and memory usage compared to the original binary, review the implications of including the C++ STL and using dynamic memory allocations for the game list.

**Comment by anzz1** [comment]
Oh wait, the functionality to use custom rom name already exists !! :open_mouth: You can do that by placing a "miyoogamelist.xml" to the roms folder. This exact functionality already works in 4.2.0.2 beta version, so no need for a new format.

As expected, it makes browsing the game list slightly more sluggish than using the default libgamename, as it needs to parse the data from a file, but its only very slightly slower and still snappy, at least for smaller libraries.

**Comment by Macs75** [comment]
> 

Maybe you missed that my library parse the file only the first time? It put it in a static variable, so the initialization of the list happens only at the first use. From the second time it uses the hashmap direct access, lot more faster that looking in the whole list and comparing each value.

**Comment by Macs75** [comment]
> 

I don't know what you tested but it's not my new lib I think, the performance are even faster for me, let me recheck the code.
Regarding the miyoogamelist.xml, yes it's true, it exist but only if you generate this XML file with external tools, so it does not impact this case. My reimplementation is for the case when this file is not present and then this function is called by the MainUi. The original lib had fixed names that could not be changed and where few. This one has a huge list of games compiled from all the supported roms by the various arcade emulators present. So a new user that just copies his romset in the miyoo will see already all the roms he has with correct names without further configuration.
I also suggest you to try the new tool in Apps that actually reduce the txt file to only the roms present on the system. This speeds up further even the first parse. It gives also a convenient output file with the missing roms he does not know that can be edited an set manually in the file.

**Pull request closed without merging**